### PR TITLE
[gentoo-build] Add --update command line argument to gentoo-build

### PR DESCRIPTION
**Pull request by vojeroen**
Split up from #14

### CODE REVIEW

File: binary_gentoo/internal/cli/build.py [hunk 1]
```diff
@@ -144,10 +149,21 @@ def build(config):
...
+            '--changed-use',
+            '--newuse',
+            '--deep',
+        ]
+    else:
+        emerge_args += [
+            '--keep-going',
+            '--with-bdeps=y',
+            '--complete-graph',
+        ]
```
Comment (hartwork) [inline binary_gentoo/internal/cli/build.py:165]: I just started wondering: Why are we not adding those for `config.update` as well?

**Review by hartwork — COMMENTED** [review]
Looks great, one question below:

File: binary_gentoo/internal/cli/build.py (hunk 1 above)
Comment (vojeroen) [reply binary_gentoo/internal/cli/build.py:165]: The `--complete-graph` seems to be similar to `--deep`, and in my tests including it tended to trigger more dependency problems. That is why I removed it.

Somehow I felt that `--with-bdeps=y` didn't seem to make as much sense when updating. But let me know if you feel differently.

The `--keep-going` is indeed something that could be included as well, I can't remember a real reason to remove it. If you want, I put it in the common arguments.

Comment (hartwork) [reply binary_gentoo/internal/cli/build.py:167]: > The `--complete-graph` seems to be similar to `--deep`, and in my tests including it tended to trigger more dependency problems. That is why I removed it.
>
> Somehow I felt that `--with-bdeps=y` didn't seem to make as much sense when updating. But let me know if you feel differently.

My main point is that from a user point of view, _losing_ `--complete-graph y` and  `--with-bdeps=y` when _adding_ ` --update` to `gentoo-build` seems unexpected. Sentence "using --with-bdeps=y together with --complete-graph makes the graph as complete as possible" from `man emerge` of portage 3.0.20-r6 is the main reason why I mostly use both arguments everywhere, but I'm no expert on these options, really.

> The `--keep-going` is indeed something that could be included as well, I can't remember a real reason to remove it. If you want, I put it in the common arguments.

Sounds good, let's use it for both cases. Thank you.

Comment (hartwork) [reply binary_gentoo/internal/cli/build.py:167]: PS: Should we have a closer look at the concrete conflicts you'd get with `--update [..] --with-bdeps=y --complete-graph`?

Comment (vojeroen) [reply binary_gentoo/internal/cli/build.py:167]: I found something rather curious. Let me show you. I have run two identical commands, with different `emerge_args`.

**Run 1, with complete-graph**

```
    emerge_args = [
        '--oneshot',
        '--verbose',
        # '--tree',
        '--jobs=2',
        f'--load-average={cpu_threads_to_use}',
        '--buildpkg=y',
        '--pretend',
    ]
    if config.update:
        emerge_args += [
            '--update',
            '--changed-use',
            '--newuse',
            '--keep-going',
            '--with-bdeps=y',
            '--complete-graph',
        ]
    else:
        emerge_args += [
            '--keep-going',
            '--with-bdeps=y',
            '--complete-graph',
        ]
```

```
$ gentoo-build --gentoo-profile "default/linux/amd64/17.1" --etc-portage ./test-portage/ --install --update @world
[...]
These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild     U  ] sys-process/procps-3.3.17-r1:0/8::gentoo [3.3.17:0/8::gentoo] USE="kill ncurses nls (split-usr) (unicode) -elogind -modern-top (-selinux) -static-libs -systemd -test" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild   R    ] sys-apps/busybox-1.32.1-r1::gentoo  USE="static -debug -ipv6* -livecd -make-symlinks -math -mdev -pam -savedconfig (-selinux) -sep-usr -syslog -systemd" 0 KiB
[ebuild     U  ] app-arch/xz-utils-5.2.5-r1::gentoo [5.2.5::gentoo] USE="extra-filters nls (split-usr) -static-libs (-threads%*)" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] sys-apps/file-5.40-r3::gentoo [5.40-r2::gentoo] USE="bzip2 seccomp zlib -lzma -python -static-libs" ABI_X86="(64) -32 (-x32)" PYTHON_TARGETS="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild   R    ] sys-apps/net-tools-2.10::gentoo  USE="arp hostname nls -ipv6* -nis -plipconfig (-selinux) -slattach -static" 0 KiB
[ebuild   R    ] net-misc/rsync-3.2.3-r4::gentoo  USE="acl iconv ssl xattr -examples -ipv6* -lz4 -stunnel -system-zlib -xxhash -zstd" 0 KiB
[ebuild     U  ] sys-apps/man-pages-5.11-r1::gentoo [5.11::gentoo] L10N="nl* -de -fr -it -ja -pl -pt-BR -ro -ru -zh-CN" 0 KiB
[ebuild  N     ] app-i18n/man-pages-l10n-4.2.0-r1::gentoo  L10N="nl -de -fr -pl -pt-BR -ro" 0 KiB
[ebuild     U  ] sys-apps/util-linux-2.36.2-r1::gentoo [2.36.2::gentoo] USE="cramfs logger ncurses nls pam readline (split-usr) suid (unicode) -audit -build -caps -cryptsetup -fdformat -hardlink -kill -magic -python (-selinux) -slang -static-libs -su -systemd -test -tty-helpers -udev" ABI_X86="(64) -32 (-x32)" PYTHON_TARGETS="python3_9 -python3_8" 0 KiB
[ebuild   R    ] sys-process/psmisc-23.4-r1::gentoo  USE="nls -X -ipv6* (-selinux)" 0 KiB
[ebuild   R    ] net-misc/wget-1.21.1::gentoo  USE="nls pcre (ssl) zlib -cookie_check -debug -gnutls -idn -ipv6* -metalink -ntlm -static -test -uuid" 0 KiB
[ebuild     U  ] sys-apps/shadow-4.9-r1::gentoo [4.8.1-r3::gentoo] USE="acl nls pam (split-usr) su xattr -audit -bcrypt -cracklib (-selinux) -skey" 1589 KiB
[ebuild     U  ] sys-devel/gcc-10.3.0-r2:10::gentoo [10.3.0:10::gentoo] USE="(cxx) fortran (multilib) nls nptl openmp pch (pie) sanitize ssp (-ada) -d -debug -doc (-fixed-point) -go -graphite (-hardened) -jit (-libssp) -lto -objc -objc++ -objc-gc -pgo -systemtap -test -vanilla -vtv -zstd" 0 KiB
[ebuild   R    ] sys-apps/iproute2-5.10.0-r1::gentoo  USE="iptables -atm -berkdb* -caps -elf -ipv6* -libbsd -minimal (-selinux)" 0 KiB
[ebuild   R    ] net-misc/iputils-20210202::gentoo  USE="arping filecaps* nls ssl -caps -clockdiff -doc -gcrypt -idn -ipv6* -nettle -rarpd -rdisc -static -tftpd -tracepath -traceroute6" 0 KiB

Total: 15 packages (7 upgrades, 1 new, 7 reinstalls), Size of downloads: 1589 KiB
```

**Run 2, with deep**
```
    emerge_args = [
        '--oneshot',
        '--verbose',
        # '--tree',
        '--jobs=2',
        f'--load-average={cpu_threads_to_use}',
        '--buildpkg=y',
        '--pretend',
    ]
    if config.update:
        emerge_args += [
            '--update',
            '--changed-use',
            '--newuse',
            '--deep',
        ]
    else:
        emerge_args += [
            '--keep-going',
            '--with-bdeps=y',
            '--complete-graph',
        ]
```

```
$ gentoo-build --gentoo-profile "default/linux/amd64/17.1" --etc-portage ./test-portage/ --install --update @world
[...]
These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild     U  ] sys-libs/glibc-2.33-r1:2.2::gentoo [2.33:2.2::gentoo] USE="(crypt) multiarch (multilib) ssp (static-libs) -audit -caps (-cet) -compile-locales -custom-cflags -doc -gd -headers-only -multilib-bootstrap -nscd -profile (-selinux) -static-pie -suid -systemtap -test (-vanilla)" 0 KiB
[ebuild   R    ] sys-libs/ncurses-6.2-r1:0/6::gentoo  USE="cxx (split-usr) (tinfo) (unicode) -ada -debug -doc -gpm -minimal -profile -static-libs -test -threads -trace" ABI_X86="32* (64) (-x32)" 0 KiB
[ebuild   R    ] dev-libs/libffi-3.3-r2:0/7::gentoo  USE="-debug -pax-kernel% -pax_kernel -static-libs -test" ABI_X86="32* (64) (-x32)" 0 KiB
[ebuild  N     ] dev-libs/ell-0.41::gentoo  USE="-glib -pie -test" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] acct-group/messagebus-0-r1::gentoo  0 KiB
[ebuild     U  ] sys-libs/timezone-data-2021a-r1::gentoo [2021a::gentoo] USE="nls -leaps-timezone -zic-slim" 0 KiB
[ebuild  N     ] acct-user/messagebus-0-r1::gentoo  0 KiB
[ebuild     U  ] sys-libs/readline-8.1_p1-r1:0/8::gentoo [8.1_p1:0/8::gentoo] USE="(split-usr) (unicode) -static-libs -utils" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] sys-process/procps-3.3.17-r1:0/8::gentoo [3.3.17:0/8::gentoo] USE="kill ncurses nls (split-usr) (unicode) -elogind -modern-top (-selinux) -static-libs -systemd -test" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] net-libs/nghttp2-1.43.0-r2:0/1.14::gentoo [1.41.0-r2:0/1.14::gentoo] USE="threads -cxx -debug -hpack-tools -jemalloc -static-libs -test -utils -xml" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild   R    ] sys-apps/busybox-1.32.1-r1::gentoo  USE="static -debug -ipv6* -livecd -make-symlinks -math -mdev -pam -savedconfig (-selinux) -sep-usr -syslog -systemd" 0 KiB
[ebuild     U  ] app-misc/pax-utils-1.3.2::gentoo [1.3.1::gentoo] USE="seccomp -caps -debug -python" PYTHON_SINGLE_TARGET="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild     U  ] app-arch/xz-utils-5.2.5-r1::gentoo [5.2.5::gentoo] USE="extra-filters nls (split-usr) -static-libs (-threads%*)" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild   R    ] sys-libs/zlib-1.2.11-r4:0/1::gentoo  USE="(split-usr) -minizip -static-libs" ABI_X86="32* (64) (-x32)" 0 KiB
[ebuild     U  ] dev-libs/libpcre-8.45:3::gentoo [8.44:3::gentoo] USE="bzip2 cxx jit readline recursion-limit (split-usr) (unicode) zlib -libedit -pcre16 -pcre32 -static-libs*" ABI_X86="(64) -32 (-x32)" 1542 KiB
[ebuild     U  ] sys-apps/file-5.40-r3::gentoo [5.40-r2::gentoo] USE="bzip2 seccomp zlib -lzma -python -static-libs" ABI_X86="(64) -32 (-x32)" PYTHON_TARGETS="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild   R    ] sys-apps/net-tools-2.10::gentoo  USE="arp hostname nls -ipv6* -nis -plipconfig (-selinux) -slattach -static" 0 KiB
[ebuild  N     ] dev-libs/gobject-introspection-common-1.68.0::gentoo  0 KiB
[ebuild   R    ] net-libs/libtirpc-1.3.2:0/3::gentoo  USE="(split-usr) -ipv6* -kerberos -static-libs" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] app-arch/zstd-1.5.0:0/1::gentoo [1.4.9:0/1::gentoo] USE="threads -lz4 -static-libs" ABI_X86="32* (64) (-x32)" 0 KiB
[ebuild   R    ] app-admin/eselect-1.4.17::gentoo  USE="vim-syntax* -doc -emacs" 0 KiB
[ebuild     U  ] dev-lang/perl-5.34.0:0/5.34::gentoo [5.32.1:0/5.32::gentoo] USE="gdbm -berkdb* -debug -doc -ithreads -minimal" 0 KiB
[ebuild     U  ] perl-core/File-Temp-0.231.100::gentoo [0.230.900::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-File-Temp-0.231.100::gentoo [0.230.900::gentoo] 0 KiB
[ebuild     U  ] dev-libs/openssl-1.1.1k-r1:0/1.1::gentoo [1.1.1k:0/1.1::gentoo] USE="asm -bindist* -rfc3779 -sctp -sslv3 -static-libs -test -tls-compression% -tls-heartbeat -vanilla (-zlib%*)" ABI_X86="(64) -32 (-x32)" CPU_FLAGS_X86="(sse2)" 0 KiB
[ebuild   R    ] net-misc/rsync-3.2.3-r4::gentoo  USE="acl iconv ssl xattr -examples -ipv6* -lz4 -stunnel -system-zlib -xxhash -zstd" 0 KiB
[ebuild     U  ] virtual/perl-ExtUtils-MakeMaker-7.620.0::gentoo [7.440.0::gentoo] 0 KiB
[ebuild  rR    ] virtual/perl-CPAN-Meta-YAML-0.18.0-r7::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-Parse-CPAN-Meta-2.150.10-r5::gentoo  0 KiB
[ebuild     U  ] virtual/perl-Encode-3.80.0::gentoo [3.60.0::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-Data-Dumper-2.179.0::gentoo [2.174.10_rc::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-File-Spec-3.800.0::gentoo [3.780.0-r1::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-Test-Harness-3.430.0::gentoo [3.420.0-r3::gentoo] 0 KiB
[ebuild  rR    ] dev-perl/MIME-Charset-1.12.2::gentoo  L10N="-ja -zh" 0 KiB
[ebuild  rR    ] dev-perl/Pod-Parser-1.630.0-r1::gentoo  USE="-test" 0 KiB
[ebuild  rR    ] dev-perl/TermReadKey-2.370.0::gentoo  USE="-examples" 0 KiB
[ebuild  rR    ] dev-perl/Text-CharWidth-0.40.0-r1::gentoo  0 KiB
[ebuild  rR    ] dev-perl/Text-WrapI18N-0.60.0-r1::gentoo  0 KiB
[ebuild  rR    ] dev-perl/Unicode-LineBreak-2019.1.0::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-Module-Metadata-1.0.37-r1::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-Perl-OSType-1.10.0-r5::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-Scalar-List-Utils-1.550.0-r1::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-Text-ParseWords-3.300.0-r8::gentoo  0 KiB
[ebuild  rR    ] virtual/perl-podlators-4.140.0-r2::gentoo  0 KiB
[ebuild     U  ] virtual/perl-ExtUtils-CBuilder-0.280.236::gentoo [0.280.234::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-ExtUtils-Install-2.200.0::gentoo [2.140.0-r3::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-ExtUtils-Manifest-1.730.0::gentoo [1.720.0-r1::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-ExtUtils-ParseXS-3.430.0::gentoo [3.400.0-r1::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-Getopt-Long-2.520.0::gentoo [2.510.0::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-version-0.992.800::gentoo [0.992.400-r1::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-Carp-1.520.0-r1::gentoo [1.500.0-r3::gentoo] 0 KiB
[ebuild     U  ] virtual/perl-Exporter-5.760.0::gentoo [5.740.0::gentoo] 0 KiB
[ebuild  rR    ] dev-perl/YAML-Tiny-1.730.0::gentoo  USE="-minimal -test" 0 KiB
[ebuild     U  ] virtual/perl-JSON-PP-4.60.0::gentoo [4.40.0::gentoo] 0 KiB
[ebuild  rR    ] virtual/perl-CPAN-Meta-2.150.10-r5::gentoo  0 KiB
[ebuild  rR    ] dev-perl/Module-Build-0.422.400::gentoo  USE="-test" 0 KiB
[ebuild  rR    ] dev-perl/SGMLSpm-1.1-r1::gentoo  0 KiB
[ebuild   R    ] net-firewall/iptables-1.8.7:0/1.8.3::gentoo  USE="(split-usr) -conntrack -ipv6* -netlink -nftables -pcap -static-libs" 0 KiB
[ebuild  N     ] app-eselect/eselect-vi-1.2::gentoo  0 KiB
[ebuild   R    ] dev-libs/expat-2.4.1::gentoo  USE="(split-usr) unicode -examples -static-libs" ABI_X86="32* (64) (-x32)" 0 KiB
[ebuild     U  ] dev-db/sqlite-3.35.5:3::gentoo [3.34.1:3::gentoo] USE="readline -debug -doc -icu -secure-delete -static-libs -tcl -test -tools" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild   R    ] net-misc/curl-7.77.0-r1::gentoo  USE="ftp http2 imap openssl pop3 progress-meter smtp ssl tftp -adns -alt-svc -brotli -gnutls -gopher -hsts -idn -ipv6* -kerberos -ldap -mbedtls -metalink (-nghttp3) -nss (-quiche) -rtmp -samba -ssh -sslv3 -static-libs -telnet -test -threads (-winssl) -zstd" ABI_X86="(64) -32 (-x32)" CURL_SSL="openssl -gnutls -mbedtls -nss (-winssl)" 0 KiB
[ebuild  rR    ] dev-perl/XML-Parser-2.440.0::gentoo  0 KiB
[ebuild  N     ] dev-libs/libuv-1.41.1:0/1::gentoo  ABI_X86="(64) -32 (-x32)" 1252 KiB
[ebuild   R    ] dev-libs/nettle-3.7.2:0/8-6::gentoo  USE="asm gmp -doc -static-libs -test" ABI_X86="(64) -32 (-x32)" CPU_FLAGS_X86="aes* -sha" 0 KiB
[ebuild  N     ] app-editors/vim-core-8.2.0814::gentoo  USE="acl nls -minimal" 0 KiB
[ebuild  NS    ] app-text/docbook-xml-dtd-4.4-r3:4.4::gentoo [4.1.2-r7:4.1.2::gentoo, 4.2-r3:4.2::gentoo, 4.5-r2:4.5::gentoo] 0 KiB
[ebuild   R    ] sys-apps/iproute2-5.10.0-r1::gentoo  USE="iptables -atm -berkdb* -caps -elf -ipv6* -libbsd -minimal (-selinux)" 0 KiB
[ebuild  N     ] app-editors/vim-8.2.0814-r100::gentoo  USE="acl nls -X -cscope -debug -gpm -ipv6 -lua -minimal -perl -python -racket -ruby (-selinux) -sound -tcl -terminal -vim-pager" LUA_SINGLE_TARGET="lua5-1 -luajit" PYTHON_SINGLE_TARGET="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild  N     ] app-vim/eselect-syntax-20070506-r1::gentoo  0 KiB
[ebuild  N     ] app-vim/gentoo-syntax-20210428::gentoo  USE="-ignore-glep31" 0 KiB
[ebuild   R    ] sys-libs/pam-1.5.1::gentoo  USE="filecaps* (split-usr) -audit -berkdb* -debug -nis (-selinux)" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] sys-devel/gcc-10.3.0-r2:10::gentoo [10.3.0:10::gentoo] USE="(cxx) fortran (multilib) nls nptl openmp pch (pie) sanitize ssp (-ada) -d -debug -doc (-fixed-point) -go -graphite (-hardened) -jit (-libssp) -lto -objc -objc++ -objc-gc -pgo -systemtap -test -vanilla -vtv -zstd" 0 KiB
[ebuild   R    ] sys-process/psmisc-23.4-r1::gentoo  USE="nls -X -ipv6* (-selinux)" 0 KiB
[ebuild   R    ] net-misc/wget-1.21.1::gentoo  USE="nls pcre (ssl) zlib -cookie_check -debug -gnutls -idn -ipv6* -metalink -ntlm -static -test -uuid" 0 KiB
[ebuild  rR    ] dev-perl/Locale-gettext-1.70.0::gentoo  0 KiB
[ebuild  N     ] app-crypt/rhash-1.3.6-r1::gentoo  USE="nls ssl -debug -static-libs" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] dev-libs/elfutils-0.185::gentoo  USE="bzip2 nls utils -lzma -static-libs -test (-threads) -valgrind -zstd" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild     U  ] app-editors/nano-5.6.1-r2::gentoo [5.6.1::gentoo] USE="ncurses nls spell (split-usr) (unicode) -debug -justify -magic -minimal -static" 0 KiB
[ebuild     U  ] sys-apps/util-linux-2.36.2-r1::gentoo [2.36.2::gentoo] USE="cramfs logger ncurses nls pam readline (split-usr) suid (unicode) -audit -build -caps -cryptsetup -fdformat -hardlink -kill -magic -python (-selinux) -slang -static-libs -su -systemd -test -tty-helpers -udev" ABI_X86="(64) -32 (-x32)" PYTHON_TARGETS="python3_9 -python3_8" 0 KiB
[ebuild     U  ] app-text/po4a-0.57-r1::gentoo [0.57::gentoo] USE="-test" 0 KiB
[ebuild  N     ] virtual/libelf-3:0/1::gentoo  ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild   R    ] sys-apps/man-db-2.9.4::gentoo  USE="gdbm manpager nls seccomp zlib -berkdb* (-selinux) -static-libs" 0 KiB
[ebuild  N     ] app-text/xmlto-0.0.28-r6::gentoo  USE="-latex -text" 0 KiB
[ebuild  N     ] sys-apps/dbus-1.12.20-r1::gentoo  USE="-X -debug -doc -elogind (-selinux) -static-libs -systemd -test -user-session" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] app-i18n/man-pages-l10n-4.2.0-r1::gentoo  L10N="nl -de -fr -pl -pt-BR -ro" 0 KiB
[ebuild     U  ] sys-apps/man-pages-5.11-r1::gentoo [5.11::gentoo] L10N="nl* -de -fr -it -ja -pl -pt-BR -ro -ru -zh-CN" 0 KiB
[ebuild     U  ] sys-apps/shadow-4.9-r1::gentoo [4.8.1-r3::gentoo] USE="acl nls pam (split-usr) su xattr -audit -bcrypt -cracklib (-selinux) -skey" 1589 KiB
[ebuild   R    ] net-misc/openssh-8.6_p1-r2::gentoo  USE="pam pie scp ssl -X -X509 -audit -bindist* (-debug) -hpn -kerberos -ldns -libedit -livecd -sctp -security-key (-selinux) -static -test -xmss" 0 KiB
[ebuild  N     ] app-arch/libarchive-3.5.1-r1:0/13::gentoo  USE="acl bzip2 e2fsprogs iconv lzma xattr zlib -blake2 -expat -lz4 -lzo -nettle -static-libs -zstd" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] dev-libs/icu-69.1:0/69.1::gentoo  USE="-debug -doc -examples -static-libs" ABI_X86="32 (64) (-x32)" 0 KiB
[ebuild  N     ] dev-libs/jsoncpp-1.9.4:0/24::gentoo  USE="-doc -test" 0 KiB
[ebuild  N     ] dev-util/cmake-3.18.5::gentoo  USE="ncurses -doc -emacs -qt5 -test" 0 KiB
[ebuild  N     ] dev-libs/json-c-0.15:0/5::gentoo  USE="-doc -static-libs -threads" ABI_X86="(64) -32 (-x32)" CPU_FLAGS_X86="-rdrand" 0 KiB
[ebuild   R    ] dev-libs/libxml2-2.9.12-r3:2::gentoo  USE="python readline -debug -examples -icu -ipv6* -lzma -static-libs -test -verify-sig" ABI_X86="32* (64) (-x32)" PYTHON_TARGETS="python3_9 -python3_8" 0 KiB
[ebuild  N     ] dev-util/itstool-2.0.6-r1::gentoo  PYTHON_SINGLE_TARGET="python3_9 -python3_8" 0 KiB
[ebuild     U  ] dev-util/meson-0.58.1::gentoo [0.57.2::gentoo] USE="(-test)" PYTHON_TARGETS="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild   R    ] dev-util/ninja-1.10.1::gentoo  USE="vim-syntax* -doc -emacs -test" 0 KiB
[ebuild   R    ] net-misc/iputils-20210202::gentoo  USE="arping filecaps* nls ssl -caps -clockdiff -doc -gcrypt -idn -ipv6* -nettle -rarpd -rdisc -static -tftpd -tracepath -traceroute6" 0 KiB
[ebuild  N     ] dev-libs/glib-2.68.3-r1:2::gentoo  USE="elf mime xattr -dbus -debug -fam -gtk-doc (-selinux) -static-libs -sysprof -systemtap -test -utils" ABI_X86="(64) -32 (-x32)" 0 KiB
[ebuild  N     ] x11-misc/shared-mime-info-2.0-r2::gentoo  0 KiB
[ebuild  N     ] dev-util/desktop-file-utils-0.26-r1::gentoo  USE="-emacs" 0 KiB
[ebuild  N     ] dev-libs/gobject-introspection-1.68.0::gentoo  USE="-doctool -gtk-doc -test" PYTHON_SINGLE_TARGET="python3_9 (-python3_10) -python3_8" 0 KiB
[ebuild  N     ] dev-libs/libical-3.0.10:0/3::gentoo  USE="glib introspection -berkdb -doc -examples -static-libs -test -vala" 0 KiB
[ebuild  N     ] net-wireless/bluez-5.58-r1:0/3::gentoo  USE="mesh obex readline udev -btpclient -cups -debug -deprecated -doc -experimental -extra-tools -midi (-selinux) -systemd -test -test-programs -user-session" ABI_X86="(64) -32 (-x32)" PYTHON_SINGLE_TARGET="python3_9 -python3_8" 0 KiB
[ebuild   R    ] dev-lang/python-3.9.5_p2:3.9::gentoo  USE="bluetooth* gdbm ncurses readline sqlite ssl xml -build -examples -hardened -ipv6* -test -tk -verify-sig -wininst" 0 KiB
[ebuild  NS    ] dev-lang/python-3.8.10_p2:3.8::gentoo [3.9.5_p2:3.9::gentoo] USE="bluetooth* gdbm ncurses readline sqlite ssl xml -build -examples -hardened -ipv6* -test -tk -verify-sig -wininst" 0 KiB
[ebuild   R    ] dev-python/certifi-10001-r1::gentoo  USE="-test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/setuptools-57.0.0::gentoo  USE="-test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/toml-0.10.2::gentoo  USE="-test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/setuptools_scm-6.0.1-r1::gentoo  USE="-test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/chardet-4.0.0::gentoo  USE="-test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild     U  ] dev-python/idna-3.2::gentoo [3.1::gentoo] PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/PySocks-1.7.1-r1::gentoo  PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild     U  ] dev-python/urllib3-1.26.5::gentoo [1.26.4::gentoo] USE="-brotli -test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB
[ebuild   R    ] dev-python/requests-2.25.1-r2::gentoo  USE="-socks5 -test" PYTHON_TARGETS="python3_8* python3_9 (-pypy3) (-python3_10)" 0 KiB

Total: 116 packages (38 upgrades, 28 new, 2 in new slots, 48 reinstalls), Size of downloads: 4383 KiB

The following packages are causing rebuilds:

  (dev-lang/perl-5.34.0:0/5.34::gentoo, ebuild scheduled for merge) causes rebuilds for:
    (dev-perl/Text-WrapI18N-0.60.0-r1:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/Unicode-LineBreak-2019.1.0:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-CPAN-Meta-YAML-0.18.0-r7:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/TermReadKey-2.370.0:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-Parse-CPAN-Meta-2.150.10-r5:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/Pod-Parser-1.630.0-r1:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/YAML-Tiny-1.730.0:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/Locale-gettext-1.70.0:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/SGMLSpm-1.1-r1:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-CPAN-Meta-2.150.10-r5:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/Module-Build-0.422.400:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/XML-Parser-2.440.0:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-Text-ParseWords-3.300.0-r8:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-Module-Metadata-1.0.37-r1:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-podlators-4.140.0-r2:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/Text-CharWidth-0.40.0-r1:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-Perl-OSType-1.10.0-r5:0/0::gentoo, ebuild scheduled for merge)
    (dev-perl/MIME-Charset-1.12.2:0/0::gentoo, ebuild scheduled for merge)
    (virtual/perl-Scalar-List-Utils-1.550.0-r1:0/0::gentoo, ebuild scheduled for merge)
```

So in the end, no difference in dependency conflicts, but it does seem that `--deep` does a more thorough job in updating packages (including the whole Perl upgrade) than `--with-bdeps=y --complete-graph`.

I also tried with all three `--deep --with-bdeps=y --complete-graph` but that basically gave the same result as with only `--deep`.

I haven't tested (yet) for the more difficult desktop/plasma profile.

Comment (vojeroen) [reply binary_gentoo/internal/cli/build.py:167]: Similar test with the profile "default/linux/amd64/17.1/desktop/plasma" yields the following results:

With `--deep`: Total: 191 packages (5 upgrades, 168 new, 18 reinstalls, 183 binaries).

Then when working with `--deep --with-bdeps=y', a lot more packages were pulled in. Total: 263 packages (15 upgrades, 205 new, 3 in new slots, 40 reinstalls, 254 binaries). This makes sense of course, since build deps are included. If we want to be as complete as possible, then probably a good idea to include this flag.

With `--complete-graph', there are less packages installed. Total: 165 packages (162 new, 3 reinstalls, 160 binaries).

With only `--with-bdeps=y --complete-graph`, it has again increased. Total: 204 packages (195 new, 3 in new slots, 6 reinstalls, 198 binaries).

In all cases, no dependency problems.

So to summarize:
* `--deep` pulls in more packages than `--complete-graph`, so it seems to upgrade more packages. No idea why, I'm not that familiar with portage. So let's use `--deep` instead of `--complete-graph` for updates?
* `--with-bdeps=y` seems a good idea to include, to be as complete as possible. Otherwise the created image will not be consistent (including bdeps when compilation is necessary, otherwise not).

Comment (hartwork) [reply binary_gentoo/internal/cli/build.py:167]: @vojeroen I think we may have a misunderstanding here. `man emerge` has a sentence that confirms well what you are experiencing above:

> Note that, unlike the `--deep` option, the `--complete-graph` option does not cause any more packages to be updated than would have otherwise been updated with the option disabled.

So let's add `--deep`, yes, no doubt about that.

The question I think we need to answer is whether `gentoo-build --update [..]` should pass…

- a) `--update --deep --complete-graph` or
- b) `--update --deep`

… and why. If (a) is causing trouble, I'd like to understand which trouble.

Comment (vojeroen) [reply binary_gentoo/internal/cli/build.py:167]: Option (a) is not really causing trouble, so we can go that way if you prefer. I'm just not sure if including `--complete-graph` if `--deep` is already there, has any additional value. But it doesn't seem to hurt either.

Comment (hartwork) [reply binary_gentoo/internal/cli/build.py:167]: Cool, let's try this then?
```
emerge_args = [
    '--oneshot',
    '--verbose',
    '--tree',
    '--jobs=2',
    f'--load-average={cpu_threads_to_use}',
    '--buildpkg=y',
    '--keep-going',
    '--with-bdeps=y',
    '--complete-graph',
]

if config.update:
    emerge_args += [
        '--update',
        '--changed-use',
        '--newuse',
        '--deep',
    ]
```

File: binary_gentoo/internal/cli/build.py [hunk 2]
```diff
@@ -146,10 +151,18 @@ def build(config):
         '--jobs=2',
         f'--load-average={cpu_threads_to_use}',
         '--buildpkg=y',
-        '--keep-going',
         '--with-bdeps=y',
+        '--keep-going',
```
Comment (hartwork) [inline binary_gentoo/internal/cli/build.py:155]: Can we revert this flip in order for a clean diff? Thank you!

**Review by hartwork — APPROVED** [review]
